sdk/core/azure-core/azure/core/pipeline/policies/_universal.py

```diff
@@ -140,9 +147,21 @@ def on_request(self, request):
         :param request: The PipelineRequest object
         :type request: ~azure.core.pipeline.PipelineRequest
         """
-        request_id = request.context.options.pop('request_id', self._request_id)
-        if request_id:
-            header = {"x-ms-client-request-id":request_id}
+        update_header = False
+        if 'request_id' in request.context.options:
+            request_id = request.context.options.pop('request_id')
+            update_header = True
+        elif self._request_id:
+            request_id = self._request_id
+            update_header = True
+        elif self._no_request_id:
+            request_id = None
+            update_header = True
+        elif self._auto_request_id:
+            request_id = str(uuid.uuid1())
+            update_header = True
+        if update_header:
+            header = {"x-ms-client-request-id": request_id}
             request.http_request.headers.update(header)
```

Comment:
separate `update_header` is not necessary:
```
if request_id is not None:
    header = {"x-ms-client-request-id": request_id}
    request.http_request.headers.update(header)
```

Reply:
I'm afraid this cannot handle the case that user explicitly sets request_id to None correctly. (our expected behavior is if user does not want to send request_id, they can explicitly set it to None.

Reply:
Then I don't understand this block:
```
        elif self._no_request_id:
             request_id = None
             update_header = True
```
which was added. That will result in a header being sent with a value of `None`

Reply:
If self._no_request_id (user explicitly opt-out to send request_id), we will clear existing request_id in the header.

Reply:
It's possible I don't understand the requirements. If so, please ignore. But my current suggestion is that if `None` is a meaningful value for users then the best thing is to explicitly distinguish it from "unset":

```
class _Unset: pass
```

```
self._request_id = kwargs.pop('request_id', _Unset)
```

``` 
         # user specifying None explicitly takes precedence (send nothing)
         if self._request_id is None:
            request_id = None
         
         # context options take next highest precendence
         elif 'request_id' in request.context.options:
             request_id = request.context.options.pop('request_id')

         # explicit user-specified next
         elif self._request_id is not _Unset:
             request_id = self._request_id
         
         # otherwise, auto-generate (if asked)
         elif self._auto_request_id:
             request_id = str(uuid.uuid1())

         $ finally, send nothing then
         else: 
            request_id = None

         if request_id is not None:
             header = {"x-ms-client-request-id": request_id}
             request.http_request.headers.update(header)

```

Reply:
The strategy is 
P1. If user has request_id with the call (request.context.options contains 'request_id'), we use it.
P2. If user called set_request_id or have a request_id when initialize RequestIdPolicy (if self._request_id), we use it
P3. If user explicitly set request_id to None (if not request_id and no_request_id), we clear existing request id
P4. If user does not specify request_id but auto_request_id is ON, we generate a unique id
P5. If user does not specify request_id and auto_request_id is OFF, we do nothing

Reply:
> clear existing request_id in the header.

This is the bit of context I was missing, thank you!

Reply:
Yes. This is tricky. :)